service/src/database/dips.rs

```
@@ -0,0 +1,130 @@
+        self.data.try_write()?.remove(&id);
+
+        Ok(id)
+    }
+}
+
+#[derive(Debug)]
+pub struct PsqlAgreementStore {
+    pool: PgPool,
+}
+
+#[async_trait]
+impl AgreementStore for PsqlAgreementStore {
+    async fn get_by_id(&self, id: Uuid) -> anyhow::Result<Option<SignedIndexingAgreementVoucher>> {
+        let item = sqlx::query!(
+            "SELECT * FROM dips_agreements WHERE signature=$1",
+            id.as_bytes(),
+        )
+        .fetch_one(&self.pool)
+        .await;
+
+        let item = match item {
+            Ok(item) => item,
+            Err(sqlx::Error::RowNotFound) => return Ok(None),
+            Err(err) => bail!(err),
+        };
+
+        let signed = SignedIndexingAgreementVoucher::decode(&mut item.signed_payload.as_ref())?;
+
+        Ok(Some(signed))
+    }
+    async fn create_agreement(
+        &self,
+        id: Uuid,
+        agreement: SignedIndexingAgreementVoucher,
+        protocol: String,
+    ) -> anyhow::Result<()> {
+        let bs = agreement.encode_vec();
+        let now = Utc::now();
+
+        sqlx::query!(
+            "INSERT INTO dips_agreements VALUES ($1,$2,$3,$4,$5,$6,$7,$8,$9,null,null)",
+            id,
+            agreement.signature.as_ref(),
+            bs,
+            protocol,
+            agreement.voucher.service.as_slice(),
+            agreement.voucher.payee.as_slice(),
+            agreement.voucher.payer.as_slice(),
+            now,
+            now,
+        )
+        .execute(&self.pool)
+        .await?;
+
+        Ok(())
+    }
+    async fn cancel_agreement(
+        &self,
+        id: Uuid,
+        signed_cancellation: SignedCancellationRequest,
+    ) -> anyhow::Result<Uuid> {
+        let bs = signed_cancellation.encode_vec();
+
+        sqlx::query!(
+            "UPDATE dips_agreements SET updated_at=now(), cancelled_at=now(), signed_cancellation_payload=$1 WHERE id=$2",
+bs,
+            id,
+        )
+        .execute(&self.pool)
+        .await?;
```

I would return an error when there's no agreement in the DB with the provided ID. If I am correct, the query will not update anything and will return successfully.

---

Isn't that the correct outcome anyway? If the query doesn't update anything it's a noop. I don't have a strong preference, if you do I'll change it otherwise I think I'll leave that in for the moment